### PR TITLE
Implement charts multiplier for SNMP collector

### PR DESCRIPTION
**Pull request by codeguru1**

### CODE REVIEW

File: modules/snmp/snmp.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ type (
 		Type          *string     `yaml:"type,omitempty"`
 		Family        *string     `yaml:"family,omitempty"`
 		MultiplyRange []int       `yaml:"multiply_range,omitempty"`
```
Comment (ilyam8) [inline modules/snmp/snmp.go:70]: We don't need `omitempty` tag.

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

File: modules/snmp/charts.go [hunk 2]
```diff
@@ -32,48 +32,65 @@ var defaultDims = module.Dims{
...
-func newChart(chartIn []ChartsConfig) *module.Charts {
-	charts := &module.Charts{}
-	for i, s := range chartIn {
-		c := defaultSNMPchart.Copy()
-		c.ID = fmt.Sprintf(c.ID, i)
-		c.Title = fmt.Sprintf(c.Title, s.Title)
-		if s.Family != nil {
-			c.Fam = *s.Family
-		}
+func newChart(id, multiplier int, s ChartsConfig) *module.Chart {
```
Comment (ilyam8) [inline modules/snmp/charts.go:35]: Let's rename `multiplier` to `oidIndex`

File: modules/snmp/charts.go [hunk 3]
```diff
@@ -32,48 +32,65 @@ var defaultDims = module.Dims{
...
+	c := defaultSNMPchart.Copy()
+	c.ID = fmt.Sprintf(c.ID, id)
+	c.Title = fmt.Sprintf(c.Title, s.Title)
 
-		if s.Units != nil {
-			c.Units = *s.Units
-		}
+	if multiplier != 0 {
```
Comment (ilyam8) [inline modules/snmp/charts.go:40]: `0` is a valid oid index.

File: modules/snmp/charts.go [hunk 4]
```diff
@@ -32,48 +32,65 @@ var defaultDims = module.Dims{
...
+	}
+
+	if s.Type != nil {
+		c.Type = module.ChartType(*s.Type)
+	}
+
+	c.Priority = s.Priority
+	for _, d := range s.Dimensions {
+		oid := d.OID
+		if multiplier != 0 {
```
Comment (ilyam8) [inline modules/snmp/charts.go:60]: 0 is a valid oid index.

File: modules/snmp/charts.go [hunk 5]
```diff
@@ -32,48 +32,65 @@ var defaultDims = module.Dims{
...
+	charts := &module.Charts{}
+	for i, s := range chartIn {
+		if s.MultiplyRange != nil {
+			for j := s.MultiplyRange[0]; j <= s.MultiplyRange[1]; j++ {
+				chart := newChart(i, j, s)
+				_ = charts.Add(chart)
+			}
+		} else {
+			chart := newChart(i, 0, s)
+			_ = charts.Add(chart)
```
Comment (ilyam8) [inline modules/snmp/charts.go:92]: Let's don't ignore an error

File: modules/snmp/validators.go [hunk 6]
```diff
@@ -77,14 +77,25 @@ func (c ChartsConfig) validateConfig(index_chart int) error {
...
 	}
 
+	if c.MultiplyRange != nil {
+		if len(c.MultiplyRange) != 2 {
+			err = appendError(err, fmt.Sprintf("invalid range: charts[%d].multiply_range;", index_chart))
+		} else {
+			if c.MultiplyRange[0] >= c.MultiplyRange[1] {
+				err = appendError(err, fmt.Sprintf("invalid range: charts[%d].multiply_range;", index_chart))
+			}
+		}
```
Comment (ilyam8) [inline modules/snmp/validators.go:97]: What about the case when one or both values are negative?

File: modules/snmp/snmp.go (hunk 1 above)
Comment (codeguru1) [reply modules/snmp/snmp.go:70]: agree, fixed in separate PR

Comment (ilyam8) [reply modules/snmp/snmp.go:70]: perfect, thanks!

File: modules/snmp/snmp.go [hunk 7]
```diff
@@ -135,13 +135,21 @@ func (s *SNMP) Init() bool {
...
-		_ = c.AddDim(defaultDims[0])
-		_ = c.AddDim(defaultDims[1])
+		for _, d := range defaultDims {
+			if err = c.AddDim(d); err != nil {
+				s.Errorf("Population of charts failed: %v", err)
+				return false
+			}
+		}
 		s.charts = &module.Charts{c}
 	}
```
Comment (ilyam8) [inline modules/snmp/snmp.go:154]: I don't think we need that `else` case with the default chart. The assumption that we need to collect ifInOctets/ifOutOctets for interface with index 2 (why 2?) looks pretty questionable to me.

Comment (ilyam8) [reply modules/snmp/snmp.go:154]: I see that it is not a new code. So we can merge it as is, but let's remove this case right away in the following PR.

Comment (ilyam8) [reply modules/snmp/snmp.go:154]: Perhaps I am missing something, if so please elaborate on why you think we need that.

File: config/go.d/snmp.conf [hunk 8]
```diff
@@ -162,22 +163,19 @@ jobs:
...
         family: "lan"
-        multiply_range: [1,24]
+        multiply_range: [1,5]
         dimensions:
           - name: "in"
-            oid: "1.3.6.1.2.1.2.2.1.10.2"
+            oid: ".1.3.6.1.2.1.2.2.1.10"
             algorithm: "incremental"
             multiplier: 8
             divisor: 1024
```
Comment (ilyam8) [inline config/go.d/snmp.conf:177]: The correct divisor is 1000 when converting bits to kilobits.

File: modules/snmp/snmp.go (hunk 7 above)
Comment (ilyam8) [reply modules/snmp/snmp.go:154]: Ok, added it to https://github.com/netdata/go.d.plugin/issues/648#issuecomment-1057198748

**Review by ilyam8 — APPROVED** [review]

**Pull request merged**